### PR TITLE
[SPARK-31634][SQL] "show tables like" support for SQL wildcard characters (% and _)

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Make `SHOW TABLE LIKE` support `%` and `_`, follow other popular DBs
https://docs.snowflake.com/en/sql-reference/sql/show-tables.html
https://clickhouse.tech/docs/en/sql-reference/statements/show/
https://www.mysqltutorial.org/mysql-show-tables/
https://issues.apache.org/jira/browse/HIVE-23359

### Why are the changes needed?
Follow common standard

### Does this PR introduce _any_ user-facing change?
Yes

`SHOW TABLES LIKE 'table*' `should be replaced as  `SHOW TABLES LIKE 'table%'`

### How was this patch tested?
Existed UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 1]
```diff
@@ -81,7 +81,7 @@ object StringUtils extends Logging {
   /**
    * This utility can be used for filtering pattern in the "Like" of "Show Tables / Functions" DDL
    * @param names the names list to be filtered
-   * @param pattern the filter pattern, only '*' and '|' are allowed as wildcards, others will
+   * @param pattern the filter pattern, only '%', '_' and '|' are allowed as wildcards, others will
```
Comment (imback82) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:84]: Do we need to add a test for the new wildcard characters supported?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:84]: > Do we need to add a test for the new wildcard characters supported?

emmm, I am waiting current but test result and fix current failed unit test and add test for new wildcard.

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
Gentle ping @wangyum @HyukjinKwon @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 2]
```diff
@@ -60,10 +60,16 @@ object StringUtils extends Logging {
         case c if c == escapeChar => fail("it is not allowed to end with the escape character")
         case '_' => out ++= "."
         case '%' => out ++= ".*"
+        case '*' if filterPattern => out ++= ".*"
+        case '.' if filterPattern => out ++= "."
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:64]: The title of this PR seems to support for SQL wildcard characters (% and _) , why we added '*' and '.' ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 3]
```diff
@@ -81,16 +87,16 @@ object StringUtils extends Logging {
...
+   *                will follow regular expression convention, case insensitive match and white
+   *                spaces on both ends will be ignored
    * @return the filtered names list in order
    */
   def filterPattern(names: Seq[String], pattern: String): Seq[String] = {
     val funcNames = scala.collection.mutable.SortedSet.empty[String]
     pattern.trim().split("\\|").foreach { subPattern =>
       try {
-        val regex = ("(?i)" + subPattern.replaceAll("\\*", ".*")).r
+        val regex = ("(?i)" + escapeLikeRegex(subPattern, '\\', true)).r
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:99]: `escapeLikeRegex(subPattern.replaceAll("\\*", "%", '\\', true)` ?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:99]: > `escapeLikeRegex(subPattern.replaceAll("\\*", "%", '\\', true)` ? so we not need to change the behavior of `escapeLikeRegex`.

Goot suggestion, updated

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:97]: There are some conflict between these two logic...maybe we should follow other DB's rule.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 4]
```diff
@@ -774,9 +774,9 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       StatsSetupConst.TOTAL_SIZE).forall(!tableMetadata.contains(_)))
 
     if (analyzedByHive) {
-      assert(StringUtils.filterPattern(describeResult1, "*numRows\\s+500*").nonEmpty)
```
Comment (beliefer) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:777]: `*` will not be supported ?

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:777]: > `*` will not be supported ?

It's the problem of `\\s+` not supported

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 4 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:777]: So this PR will break change the behavior of `StringUtils.filterPattern` ?

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:777]: > So this PR will break change the behavior of `StringUtils.filterPattern` ?

We can add a little logic in `escapeLikeRegex` to solve this.

**Pull request closed without merging**